### PR TITLE
[ENH] Add sliding window utils for anomaly detection module

### PR DESCRIPTION
**Pull request by CodeLionX**
#### Reference Issues/PRs

none

#### What does this implement/fix? Explain your changes.

Adds two functions to a new utility module `aeon.utils.windowing`:

- `sliding_windows(X, window_size, stride, axis=1)` to create a sliding window view on a time series
- `reverse_windowing(y, window_size, reduction, stride, padding_length)` to reverse window anomaly scores to point anomaly scores

The windowing functions will be used in the implementation of `KMeansAnomalyDetector` and the `PyODAdapter` (#1586).

The new util functions are not added to the API reference.

#### Does your contribution introduce a new dependency? If yes, which one?

no

##### For all contributions

- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

### CODE REVIEW

**Comment by TonyBagnall** [comment]
would it be possible to add an axis argument for time dimension, default to 1?

File: pyproject.toml [hunk 1]
```diff
@@ -186,6 +186,7 @@ addopts = '''
     --dist worksteal
     --reruns 2
     --only-rerun "crashed while running"
+    --ignore-glob 'aeon/testing/utils/_*.py'
```
Comment (MatthewMiddlehurst) [inline pyproject.toml:189]: Why is this required?

File: aeon/utils/windowing.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+    # if we are running within a dask worker
+    try:
+        import dask.distributed
+
+        dask_memory_limit = dask.distributed.get_worker().memory_limit
+        if dask_memory_limit is not None and dask_memory_limit < memory_limit:
+            memory_limit = dask_memory_limit - memory_buffer
+    except (ImportError, ValueError):
+        pass
```
Comment (MatthewMiddlehurst) [inline aeon/utils/windowing.py:218]: We were looking at dropping `dask` actually. Is this necessary?

**Review by MatthewMiddlehurst — COMMENTED** [review]
Still need to have a more thorough look so will revisit unless someone else accepts, some comments though.

File: aeon/utils/windowing.py (hunk 2 above)
Comment (CodeLionX) [reply aeon/utils/windowing.py:238]: No, it is just copied from TimeEval https://github.com/TimeEval/TimeEval/blob/2a002016c93c0742e32dc68a985f56426a4d97b0/timeeval/utils/window.py#L120.
We could also replace the memory-based decision by a threshold-based decision. However, I would not know how to set the threshold to switch between the vectorized and iterative code.

Comment (MatthewMiddlehurst) [reply aeon/utils/windowing.py:238]: I think this is fine for now. Does not seem like it would cause any unintended exceptions given the try/catch.

**Review by MatthewMiddlehurst — COMMENTED** [review]
Looks good for the most part, left a few comments.

File: aeon/utils/windowing.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    >>> print(windows)
+    [[1 2 3 4]
+     [5 6 7 8]]
+    >>> print(padding)
+    2
+
+    See Also
+    --------
+    aeon.utils.windowing.reverse_windowing :
+        Reverse windowing of a time series.
```
[the remark this answers is not in the document]
Comment (CodeLionX) [reply aeon/utils/windowing.py]: Sure

**Review by MatthewMiddlehurst — APPROVED** [review]
LGTM, are you fine with the axis parameter @TonyBagnall?

**Review by TonyBagnall — APPROVED** [review]
lgtm

**Comment by TonyBagnall** [comment]
> Asked again, if we don't hear in a day or two you can just merge I would say, can deal with any issues in another PR if present.



> Asked again, if we don't hear in a day or two you can just merge I would say, can deal with any issues in another PR if present.

give me a chance :)

**Comment by MatthewMiddlehurst** [comment]
Yeah fair enough, just saw the 3 days and didnt occur to me it was the weekend straight away 🙂. Trying to get these in since theres a time limit
